drizzlepac/alignimages.py

```diff
@@ -25,646 +25,676 @@
+                    # execute the correct fitting/matching algorithm
+                    imglist = algorithm(imglist, reference_catalog)
+                except Exception:
+                    exc_type, exc_value, exc_tb = sys.exc_info()
+                    traceback.print_exception(
+                        exc_type, exc_value, exc_tb, file=sys.stdout
+                    )
+                    log.warning(
+                        "WARNING: Catastrophic fitting failure with "
+                        "catalog {} and matching algorithm {}."
+                        .format(catalog, algorithm.__name__)
+                    )
+                    fitab['status'][:] = 1
+                    fitab['processMsg'][:] = "Fitting failure"
+                    # It may be there are additional catalogs and
+                    # algorithms to try, so keep going but flag this fit
+                    # with the 'bad' quality value
                     fit_quality = 5
-                    if catalog_index < len(catalog_list) - 1:
-                        log.info("Try again with other catalog")
-                    else:
-                        # bail out if not enough sources can be found any of the astrometric catalogs
-                        log.warning("ERROR! No astrometric sources found in any catalog. Exiting...")
-                        filtered_table['status'][:] = 1
-                        filtered_table['processMsg'][:] = "No astrometric sources found"
-                        filtered_table['fit_qual'][:] = fit_quality
-                        current_dt = datetime.datetime.now()
-                        delta_dt = (current_dt - starting_dt).total_seconds()
-                        log.info('Processing time of [STEP 5]: {} sec'.format(delta_dt))
-                        return (filtered_table)
+                    fitab['fit_qual'][:] = fit_quality
+                    continue
+
+                # determine the quality of the fit
+                fit_rms, fit_num, fit_quality, fitab, fit_status = \
+                    determine_fit_quality(
+                        imglist, fitab, catalog == catalogs[-1],
+                        print_fit_parameters=print_fit_parameters
+                    )
+
+                # save fit algorithm name to dictionary key
+                # "fit method" in imglist.
+                for img in imglist:
+                    img.meta['fit method'] = algorithm.__name__
+                    img.meta['fit quality'] = fit_quality
+
+                # populate fit_info
+                kwd = "{} {}".format(
+                    catalog, algorithm.__name__
+                )
+                fit_info[kwd] = fit_status[
+                    next(iter(fit_status))
+                ]
+                fit_info[kwd]['fit_qual'] = fit_quality
+
+                # Figure out which fit solution to go with based on
+                # fit_quality value and maybe also total_rms
+                if fit_quality >= 5:
```

Comment:
This can be re-written such that `fit_quality ...: pass` is unnecessary. Also, the way this is set up right now, `best_fit_qual` is ignored when `fit_quality >= 5`.

Reply:
@pllim My understanding is that `best_fit_qual` is worst when 5 and best at 1. I was trying to avoid the situation on line 520-521 in the old file
```
if fit_quality < 5:
    if fit_quality == 1:
        ....
```
in order to decrease nesting level. If there is a better approach - let me know (or make a PR later).